lib/gc/Transforms/Pipeline.cpp

```diff
@@ -92,21 +92,27 @@ void populateVectorPasses(mlir::OpPassManager &pm) {
 
 // scf + arith + math + vector + memref + linalg.brgemm
 void populateBufferizationPasses(mlir::OpPassManager &pm) {
+  pm.addPass(bufferization::createEmptyTensorEliminationPass());
   bufferization::OneShotBufferizationOptions options;
   options.bufferizeFunctionBoundaries = true;
   options.setFunctionBoundaryTypeConversion(
       bufferization::LayoutMapOption::IdentityLayoutMap);
   pm.addPass(bufferization::createOneShotBufferizePass(options));
   pm.addPass(createCanonicalizerPass());
   pm.addPass(createCSEPass());
+  pm.addPass(createCanonicalizerPass());
```

Comment:
L101 is already CanonicalizerPass, do we really need to do it again after CSE?

Reply:
Yes, the first `CanonicalizerPass` is mainly used to canonicalize the `scf.for` and `scf.forall` to the canonical form in memref, exposing more `CSE` optimization opportunities. The `CSE` pass is used to eliminate the common `memref.subview` so there will be some useless memref.copy `memref.copy %a, %a`. The last `CanonicalizerPass` is used to eliminate the useless `memref.copy` which copies from A to A like `memref.copy %a, %a`